src/api/fetch-posts.ts

```diff
@@ -0,0 +1,32 @@
+import { fetchPostImage } from "./fetch-post-image";
+import { fetchUsername } from "./fetch-username";
+import { fetchUserImage } from "./fetch-user-image";
+
+export const fetchPosts = async () => {
+    const response = await fetch('http://localhost:3000/fetch-posts');
+
+    if (!response.ok) {
+        throw new Error(response.statusText);
+    }
+
+    const posts = await response.json();
+
+    const userPromises = posts.map(async (post: any) => {
+        const userData = await fetchUsername(post.user_id);
+        post.user = userData;
+
+        const userImage = await fetchUserImage(post.user_id);
+        post.user.image = userImage;
+    });
+
+    await Promise.all(userPromises);
+
+    const imagePromises = posts.map(async (post: any) => {
+        const image = await fetchPostImage(post.post_id);
+        post.image = image;
```

Comment:
To jest to o czym pisałem wyżej, jakby backend zwracał od razu wymagane dane to byś jednym fetchem ogarnął wszystko i nie musiał robić takich kiwek. Nawet myślę, że da się na backendzie obsłużyc fetchowanie zdjęć z firebase i zwracać dla fronta tylko image_url już z twojego backendu. To takie zadanie na plusa